editor/window_wrapper.cpp

```diff
@@ -0,0 +1,397 @@
+		ERR_FAIL_COND(!parent);
+
+		if (!parent->is_visible()) {
+			return;
+		}
+
+		// Get the window of the parent of the current window.
+		Window *grandparent_window = Object::cast_to<Window>(parent->get_parent()->get_viewport());
+		ERR_FAIL_COND(!grandparent_window);
+
+		if (Object::cast_to<InputEventKey>(p_event.ptr()) || Object::cast_to<InputEventShortcut>(p_event.ptr())) {
+			// HACK: Propagate the window input to the editor main window to handle global shortcuts.
+			grandparent_window->push_unhandled_input(p_event);
+			if (grandparent_window->is_input_handled()) {
+				get_viewport()->set_input_as_handled();
+			}
+		}
+	}
+};
+
+Rect2 WindowWrapper::_get_default_window_rect() const {
+	// Assume that the control rect is the desidered one for the window.
+	return wrapped_control->get_global_rect();
+}
+
+Node *WindowWrapper::_get_wrapped_control_parent() const {
+	if (margins) {
+		return margins;
+	}
+	return window;
+}
+
+void WindowWrapper::_bind_methods() {
+	ADD_SIGNAL(MethodInfo("window_visibility_changed", PropertyInfo(Variant::BOOL, "visible")));
+}
+
+void WindowWrapper::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_VISIBILITY_CHANGED: {
+			if (get_window_enabled() && is_visible()) {
+				window->grab_focus();
+			}
+		} break;
+		case NOTIFICATION_READY: {
+			set_process_input(true);
+		} break;
+		case NOTIFICATION_THEME_CHANGED: {
+			window_background->add_theme_style_override("panel", get_theme_stylebox("PanelForeground", "EditorStyles"));
+		} break;
+	}
+}
+
+void WindowWrapper::shortcut_input(const Ref<InputEvent> &p_event) {
+	if (enable_shortcut.is_valid() && enable_shortcut->matches_event(p_event)) {
+		set_window_enabled(true);
+	}
+}
+
+void WindowWrapper::set_wrapped_control(Control *p_control, const Ref<Shortcut> &p_enable_shortcut) {
+	ERR_FAIL_NULL(p_control);
+	ERR_FAIL_COND(wrapped_control);
+
+	wrapped_control = p_control;
+	enable_shortcut = p_enable_shortcut;
+	add_child(p_control);
+}
+
+Control *WindowWrapper::get_wrapped_control() const {
+	return wrapped_control;
+}
+
+Control *WindowWrapper::release_wrapped_control() {
+	set_window_enabled(false);
+	if (wrapped_control) {
+		Control *old_wrapped = wrapped_control;
+		remove_child(wrapped_control);
+		wrapped_control = nullptr;
+
+		return old_wrapped;
+	}
+	return nullptr;
+}
+
+bool WindowWrapper::is_window_available() const {
+	return window != nullptr;
+}
+
+void WindowWrapper::set_window_enabled(bool p_visible) {
+	ERR_FAIL_NULL(wrapped_control);
+
+	if (!is_window_available()) {
+		return;
+	}
+
+	if (window->is_visible() == p_visible) {
+		if (p_visible) {
+			window->grab_focus();
+		}
+		return;
+	}
+
+	Node *parent = _get_wrapped_control_parent();
+
+	window->set_visible(p_visible);
+	if (wrapped_control->get_parent() != parent) {
+		Rect2 control_rect = _get_default_window_rect();
+
+		// Move the control to the window.
+		remove_child(wrapped_control);
+		parent->add_child(wrapped_control);
```

Comment:
```suggestion
		wrapped_control->reparent(parent, false);
```
You can use the new `reparent()` method (applies to everywhere you reparent).